src/test/kotlin/org/opensearch/replication/task/index/IndexReplicationTaskTests.kt

```diff
@@ -0,0 +1,299 @@
+        replicationTask.setReplicationMetadata(rm)
+
+        //Update ClusterState to say restore started
+        val state: ClusterState = clusterService.state()
+
+        var newClusterState: ClusterState
+
+        // Updating cluster state
+        var builder: ClusterState.Builder
+        val indices: MutableList<String> = ArrayList()
+        indices.add(followerIndex)
+        val snapshot = Snapshot("$REMOTE_REPOSITORY_PREFIX$connectionName", SnapshotId("randomAlphaOfLength", "randomAlphaOfLength"))
+        val restoreEntry = RestoreInProgress.Entry("restoreUUID", snapshot, RestoreInProgress.State.INIT, Collections.unmodifiableList(ArrayList<String>(indices)),
+                null)
+
+        // Update metadata store index as well
+        var metaBuilder = Metadata.builder()
+        metaBuilder.put(IndexMetadata.builder(REPLICATION_CONFIG_SYSTEM_INDEX).settings(settings(Version.CURRENT)).numberOfShards(1).numberOfReplicas(0))
+        var metadata = metaBuilder.build()
+        var routingTableBuilder = RoutingTable.builder()
+        routingTableBuilder.addAsNew(metadata.index(REPLICATION_CONFIG_SYSTEM_INDEX))
+        var routingTable = routingTableBuilder.build()
+
+        builder = ClusterState.builder(state).routingTable(routingTable)
+        builder.putCustom(RestoreInProgress.TYPE, RestoreInProgress.Builder(
+                state.custom(RestoreInProgress.TYPE, RestoreInProgress.EMPTY)).add(restoreEntry).build())
+
+        newClusterState = builder.build()
+        setState(clusterService, newClusterState)
+
+        val job = this.launch{
+            replicationTask.execute(this, InitialState)
+        }
+
+        // Delay to let task execute
+        delay(1000)
+
+        // Assert we move to RESTORING .. This is blocking and won't let the test run
+        assertBusy({
+            assertThat(currentTaskState == RestoreState).isTrue()
+        },  1, TimeUnit.SECONDS)
+
+
+        //Complete the Restore
+        metaBuilder = Metadata.builder()
+        metaBuilder.put(IndexMetadata.builder(followerIndex).settings(settings(Version.CURRENT)).numberOfShards(1).numberOfReplicas(0))
+        metadata = metaBuilder.build()
+        routingTableBuilder = RoutingTable.builder()
+        routingTableBuilder.addAsNew(metadata.index(followerIndex))
+        routingTable = routingTableBuilder.build()
+
+        builder = ClusterState.builder(state).routingTable(routingTable)
+        builder.putCustom(RestoreInProgress.TYPE, RestoreInProgress.Builder(
+                state.custom(RestoreInProgress.TYPE, RestoreInProgress.EMPTY)).build())
+
+        newClusterState = builder.build()
+        setState(clusterService, newClusterState)
+
+        delay(1000)
+
+        assertBusy {
+            assertThat(currentTaskState == MonitoringState).isTrue()
+        }
+
+        job.cancel()
+
+    }
+
+    open class NoOpClient(testName :String) : NoOpNodeClient(testName) {
```

Comment:
Seems like this could be extracted out and can also be helpful for other tests.

Reply:
I have extracted out from IndexReplicationTaskTests , but kept it in the same package as it needs all variables like follower index et all and is IndexReplicationTask as of now.